synse/websocket.go

```diff
@@ -0,0 +1,515 @@
+func NewWebSocketClientV3(opts *Options) (Client, error) {
+	c, err := createWebSocketClient(opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create a websocket client")
+	}
+
+	s := "ws"
+	if opts.TLS.Enabled {
+		s = "wss"
+	}
+
+	return &websocketClient{
+		options:    opts,
+		client:     c,
+		apiVersion: "v3",
+		entryRoute: "connect",
+		scheme:     s,
+	}, nil
+}
+
+// createWebSocketClient setups a websocket dialer with configured options.
+func createWebSocketClient(opts *Options) (*websocket.Dialer, error) {
+	err := setDefaults(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	if !opts.TLS.Enabled {
+		return &websocket.Dialer{
+			HandshakeTimeout: opts.WebSocket.HandshakeTimeout,
+		}, nil
+	}
+
+	// Only return a TLS client if its config option is enable.
+	cert, err := setTLS(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &websocket.Dialer{
+		HandshakeTimeout: opts.WebSocket.HandshakeTimeout,
+		TLSClientConfig: &tls.Config{
+			Certificates: []tls.Certificate{cert},
+		},
+	}, nil
+}
+
+// Open opens the websocket connection between the client and Synse Server.
+func (c *websocketClient) Open() error {
+	conn, _, err := c.client.Dial(buildURL(c.scheme, c.options.Address, c.apiVersion, c.entryRoute), nil)
+	if err != nil {
+		return errors.Wrap(err, "failed to open the websocket connection")
+	}
+
+	c.connection = conn
+	return nil
+}
+
+// Close closes the websocket connection between the client and Synse Server.
+// It's up to the user to close the connection after finish using it.
+// FIXME - it makes sense to use defer function with Close. However, since
+// Close returns an error, do user need to check it manually? If yes, how?
+func (c *websocketClient) Close() error {
+	err := c.connection.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
+	if err != nil {
+		return errors.Wrap(err, "failed to close the connection gracefully")
+	}
+
+	return nil
+}
+
+// Status returns the status info. This is used to check if the server
+// is responsive and reachable.
+// NOTE - this method is not applicable for websocket client.
+func (c *websocketClient) Status() (*scheme.Status, error) {
+	return nil, nil
+}
+
+// Version returns the version info.
+func (c *websocketClient) Version() (*scheme.Version, error) {
+	req := scheme.RequestVersion{
+		EventMeta: scheme.EventMeta{
+			ID:    addCounter(),
+			Event: requestVersion,
+		},
+	}
+
+	resp := new(scheme.ResponseVersion)
+	err := c.makeRequest(req, resp)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.verifyResponse(req.EventMeta, resp.EventMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	// FIXME - should we return the everything from ResponseVersion
+	// or it is fine to return the Data value only since these metadata
+	// such as request id and request won't be much helpful for consumer
+	// anyway? Should the response from websocket client be similar to http
+	// client?
```

Comment:
Yes, since both the HTTP and WebSocket client fulfil the same client interface, I'd expect both of them to return the same data (e.g. I don't think we need to include the ID/Event info from the message)